app/questionnaire/router.py

```diff
@@ -1,67 +1,124 @@
+from app.helpers.path_finder_helper import path_finder
+from app.questionnaire.location import Location
+
+
 class Router:
-    def __init__(
-        self, schema, routing_path, current_location=None, completed_locations=None
-    ):
+    def __init__(self, schema, completed_items=None):
         self._schema = schema
-        self._routing_path = routing_path
-        self._current_location = current_location
-        self._completed_locations = completed_locations
+        self._completed_items = completed_items
 
-    def can_access_location(self):
+    def can_access_location(self, location: Location, routing_path):
         """
-        Checks whether the current location is valid and accessible.
+        Checks whether the location is valid and accessible.
         :return: boolean
         """
-        if (
-            self._is_current_location_valid()
-            and not self._get_survey_redirect_location()
-        ):
+        if location in routing_path:
+            block = self._schema.get_block(location.block_id)
+            if (
+                block['type'] in ['Confirmation', 'Summary']
+                and not self.is_survey_complete()
+            ):
+                return False
+
             return True
 
         return False
 
-    def get_next_location(self):
+    def get_next_location(self, location, routing_path):
         """
         Get the first incomplete block in section/survey if trying to access the section/survey end,
         and the section/survey is incomplete or gets the next default location if the above is false.
         :return: boolean
         """
-        return self._get_survey_redirect_location() or self._get_default_location()
+        current_block_type = self._schema.get_block(location.block_id)['type']
+        last_block_location = routing_path[-1]
+        last_block_type = self._schema.get_block(last_block_location.block_id)['type']
 
-    def _is_current_location_valid(self):
-        """
-        Checks whether the current location is within the routing path.
-        :return: boolean
-        """
-        return self._current_location in self._routing_path
+        # If the section is complete and contains a SectionSummary, return the SectionSummary location
+        if (
+            last_block_type == 'SectionSummary'
+            and current_block_type != last_block_type
+            and path_finder.is_path_complete(routing_path)
+        ):
+            return last_block_location
 
-    def _get_default_location(self):
-        """
-        Calculates the next location if the current location is invalid
-        or if the current location is valid but not accessible.
-        :return: Location to direct to.
-        """
+        if self.is_survey_complete():
+            all_sections = [section['id'] for section in self._schema.get_sections()]
```

Comment:
Think the var should be  exactly what the method needs. Also should be more efficient that list comp.
```python
last_section_id = list(self._schema.get_sections())[-1]['id']
```
`final_block_id` should be `last_block_id` for consistency.

Reply:
As it's done in a few places I've added a `get_section_ids` method to questionnaire_schema.py.

Reply:
And renamed `final_block_id` to `last_block_id`